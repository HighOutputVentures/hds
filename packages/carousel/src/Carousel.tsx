import {
  Box,
  Button,
  Icon,
  IconButton,
  SystemStyleObject,
  useMultiStyleConfig,
} from '@chakra-ui/react';
import * as React from 'react';
import { Carousel as ReactResponsiveCarousel } from 'react-responsive-carousel';
import CarouselStyle from './CarouselStyle.generated';
import ChevronLeftIcon from './ChevronLeftIcon';
import ChevronRightIcon from './ChevronRightIcon';

type CarouselSize = 'md' | 'lg';

type CarouselVariant = 'dot' | 'dash';

export type CarouselProps<T extends unknown[]> = SystemStyleObject & {
  items: T;
  size?: CarouselSize;
  loop?: boolean;
  variant?: CarouselVariant;
  autoPlay?: boolean;
  isSwipeable?: boolean;
  wrapIndicator?: boolean;
  children: (item: T[number], index: number) => React.ReactNode;
  __carouselContainerTestId?: string;
  __indicatorTestId?: string;
};

export default function Carousel<T extends unknown[]>({
  variant = 'dot',
  size = 'md',
  items,
  loop,
  autoPlay,
  isSwipeable,
  __indicatorTestId,
  __carouselContainerTestId,
  wrapIndicator,
  children,
  ...props
}: CarouselProps<T>) {
  const styles = useMultiStyleConfig('Carousel', {
    variant: 'hds',
    __size: size,
    __variant: variant,
    __wrapped: wrapIndicator,
  });

  return (
    <Box
      sx={{
        ...props,
        ...styles.container,
        '.control-dots': styles.dots,
      }}
      data-testid={__carouselContainerTestId ?? 'hds.carousel.container'}
    >
      <CarouselStyle />
      <ReactResponsiveCarousel
        width="100%"
        autoPlay={autoPlay}
        swipeable={isSwipeable}
        emulateTouch={isSwipeable}
        infiniteLoop={loop}
        interval={5000}
        showStatus={false}
        showThumbs={false}
        stopOnHover={false}
        renderArrowNext={(handler, _hasNext, label) => (
          <IconButton
            variant="unstyled"
            aria-label={label}
            data-testid="hds.carousel.next.btn"
            icon={<Icon as={ChevronRightIcon} />}
            onClick={handler}
            sx={styles.arrow}
          />
        )}
        renderArrowPrev={(handler, _hasPrev, label) => (
          <IconButton
            variant="unstyled"
            aria-label={label}
            data-testid="hds.carousel.previous.btn"
            icon={<Icon as={ChevronLeftIcon} />}
            onClick={handler}
            sx={styles.arrow}
          />
        )}
        renderIndicator={(handler, isSelected, index, label) => {
          return (
            <Button
              variant="unstyled"
              tabIndex={0}
              onClick={handler}
              onKeyDown={handler}
              aria-label={`${label} ${index + 1}`}
              data-selected={isSelected}
              datat-testid={__indicatorTestId ?? 'hds.carousel.indicator'}
              sx={styles.dot}
            />
          );
        }}
        labels={{
          item: 'Slide',
          leftArrow: 'previous slide',
          rightArrow: 'next slide',
        }}
      >
        {items.map((item, index) => (
          <React.Fragment key={`${index}${item}`}>{children(item, index)}</React.Fragment>
        ))}
      </ReactResponsiveCarousel>
    </Box>
  );
}
