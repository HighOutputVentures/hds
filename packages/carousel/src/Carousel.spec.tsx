import "@testing-library/jest-dom";
import { cleanup, fireEvent, render, screen } from "@testing-library/react";
import * as React from "react";
import Carousel from "./Carousel";

jest.mock("react-responsive-carousel/lib/styles/carousel.min.css", () => {
  return {};
});

describe("Carousel", () => {
  afterEach(cleanup);
  beforeEach(() => {
    const { debug } = render(
      <Carousel
        items={[
          "https://example-img-01.png",
          "https://example-img-02.png",
          "https://example-img-03.png",
          "https://example-img-04.png",
          "https://example-img-05.png",
        ]}
        loop
        autoPlay
        isSwipeable
      >
        {(src) => <img src={src} alt="" />}
      </Carousel>,
    );

    debug();
  });

  it("Should render next and previous button", () => {
    expect(screen.queryByRole("button", { name: "next slide" })).toBeDefined();
    expect(screen.queryByRole("button", { name: "previous slide" })).toBeDefined();
  });

  it("Should render first slide by default", () => {
    expect(screen.queryByRole("button", { name: "Slide 1" })).toHaveAttribute(
      "data-selected",
      "true",
    );
  });

  it("Should be able to go to next slide", () => {
    const button = screen.getByRole("button", { name: "next slide" });
    fireEvent.click(button);
    expect(screen.queryByRole("button", { name: "Slide 2" })).toHaveAttribute(
      "data-selected",
      "true",
    );
  });

  it("Should be able to go to prev slide", () => {
    const next = screen.getByRole("button", { name: "next slide" });
    const prev = screen.getByRole("button", { name: "previous slide" });

    fireEvent.click(next);
    fireEvent.click(prev);

    expect(screen.queryByRole("button", { name: "Slide 1" })).toHaveAttribute(
      "data-selected",
      "true",
    );
  });

  it("Should be able to jump to particular slide", () => {
    const button = screen.getByRole("button", { name: "Slide 3" });

    fireEvent.click(button);

    expect(screen.queryByRole("button", { name: "Slide 3" })).toHaveAttribute(
      "data-selected",
      "true",
    );
  });
});
