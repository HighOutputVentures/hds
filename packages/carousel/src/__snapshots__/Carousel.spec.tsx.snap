// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Carousel Snapshot Should match snapshot 1`] = `
<div
  className="css-0"
  data-testid="hds.carousel.container"
>
  <div
    className="carousel-root"
  >
    <div
      className="carousel carousel-slider"
      style={
        {
          "width": "100%",
        }
      }
    >
      <ul
        className="control-dots"
      >
        <button
          aria-label="Slide 1"
          className="chakra-button css-4xx2wk"
          data-selected={true}
          datat-testid="hds.carousel.indicator"
          onClick={[Function]}
          onKeyDown={[Function]}
          tabIndex={0}
          type="button"
        />
        <button
          aria-label="Slide 2"
          className="chakra-button css-4xx2wk"
          data-selected={false}
          datat-testid="hds.carousel.indicator"
          onClick={[Function]}
          onKeyDown={[Function]}
          tabIndex={0}
          type="button"
        />
        <button
          aria-label="Slide 3"
          className="chakra-button css-4xx2wk"
          data-selected={false}
          datat-testid="hds.carousel.indicator"
          onClick={[Function]}
          onKeyDown={[Function]}
          tabIndex={0}
          type="button"
        />
        <button
          aria-label="Slide 4"
          className="chakra-button css-4xx2wk"
          data-selected={false}
          datat-testid="hds.carousel.indicator"
          onClick={[Function]}
          onKeyDown={[Function]}
          tabIndex={0}
          type="button"
        />
        <button
          aria-label="Slide 5"
          className="chakra-button css-4xx2wk"
          data-selected={false}
          datat-testid="hds.carousel.indicator"
          onClick={[Function]}
          onKeyDown={[Function]}
          tabIndex={0}
          type="button"
        />
      </ul>
      <button
        aria-label="previous slide"
        className="chakra-button css-p7mkev"
        data-testid="hds.carousel.previous.btn"
        onClick={[Function]}
        type="button"
      >
        <svg
          aria-hidden={true}
          className="chakra-icon css-13otjrl"
          fill="none"
          focusable={false}
          stroke="currentColor"
          strokeWidth={1.5}
          viewBox="0 0 24 24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M15.75 19.5L8.25 12l7.5-7.5"
            strokeLinecap="round"
            strokeLinejoin="round"
          />
        </svg>
      </button>
      <div
        className="slider-wrapper axis-horizontal"
        style={{}}
      >
        <ul
          className="slider animated"
          onMouseDown={[Function]}
          onTouchEnd={[Function]}
          onTouchStart={[Function]}
          style={
            {
              "MozTransitionDuration": "350ms",
              "OTransform": "translate3d(-100%,0,0)",
              "OTransitionDuration": "350ms",
              "WebkitTransform": "translate3d(-100%,0,0)",
              "WebkitTransitionDuration": "350ms",
              "msTransform": "translate3d(-100%,0,0)",
              "msTransitionDuration": "350ms",
              "transform": "translate3d(-100%,0,0)",
              "transitionDuration": "350ms",
            }
          }
        >
          <li
            className="slide"
            onClick={[Function]}
            style={{}}
          >
            <img
              alt=""
              src="https://example-img-05.png"
            />
          </li>
          <li
            className="slide selected previous"
            onClick={[Function]}
            style={{}}
          >
            <img
              alt=""
              src="https://example-img-01.png"
            />
          </li>
          <li
            className="slide"
            onClick={[Function]}
            style={{}}
          >
            <img
              alt=""
              src="https://example-img-02.png"
            />
          </li>
          <li
            className="slide"
            onClick={[Function]}
            style={{}}
          >
            <img
              alt=""
              src="https://example-img-03.png"
            />
          </li>
          <li
            className="slide"
            onClick={[Function]}
            style={{}}
          >
            <img
              alt=""
              src="https://example-img-04.png"
            />
          </li>
          <li
            className="slide"
            onClick={[Function]}
            style={{}}
          >
            <img
              alt=""
              src="https://example-img-05.png"
            />
          </li>
          <li
            className="slide selected previous"
            onClick={[Function]}
            style={{}}
          >
            <img
              alt=""
              src="https://example-img-01.png"
            />
          </li>
        </ul>
      </div>
      <button
        aria-label="next slide"
        className="chakra-button css-p7mkev"
        data-testid="hds.carousel.next.btn"
        onClick={[Function]}
        type="button"
      >
        <svg
          aria-hidden={true}
          className="chakra-icon css-13otjrl"
          fill="none"
          focusable={false}
          stroke="currentColor"
          strokeWidth={1.5}
          viewBox="0 0 24 24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M8.25 4.5l7.5 7.5-7.5 7.5"
            strokeLinecap="round"
            strokeLinejoin="round"
          />
        </svg>
      </button>
    </div>
  </div>
</div>
`;
